Add 'config.action_text.attachment_tag_name'

markvanlan opened the pull request:
### Summary

On the [Rails forum](https://discuss.rubyonrails.org/t/what-is-the-best-way-to-change-actiontext-attachments-into-href-url-for-json-format/76955?u=markvanlan) someone wants to change actiontext's default attachment HTML tag name. This is `action-text-attachment` by default. One way to do this right now is to parse the rich_text output and change the tag name using nokogiri. This PR adds the option `config.action_text.attachment_tag_name` which changes `action-text-attachment` to whatever string is passed in.

In the unit test for `action_text/attachment`, I set `ActionText::Attachment.tag_name` and `ActionText::Attachment.selector` manually, rather than re-running a configuration to set these. Maybe there is a different way in unit tests to change app configs?

This is my first PR for Rails; please let me know how to improve it!

markvanlan committed:
Add config.action_text.attachment_tag_name

georgeclaghorn commented on actiontext/lib/action_text/railtie.rb:
```diff
@@ -0,0 +1,13 @@
+    config.action_text = ActiveSupport::OrderedOptions.new
+    config.action_text.attachment_tag_name = "action-text-attachment"
+    initializer "action_text.configure" do |app|
+      ActionText::Attachment.tag_name = app.config.action_text.attachment_tag_name
```
Need to set `ActionText::Attachment.selector` too?

georgeclaghorn commented on actiontext/lib/action_text/attachment.rb:
```diff
@@ -6,8 +6,9 @@ module ActionText
-    TAG_NAME = "action-text-attachment"
-    SELECTOR = TAG_NAME
+    mattr_accessor :tag_name, default: "action-text-attachment"
+    mattr_accessor :selector, default: tag_name
```
Is it worth keeping `selector` separate, or can we use `tag_name` everywhere we use `selector`?

markvanlan replied:
I would prefer to use tag_name everywhere. I don't see any use keeping them separate if they will never be different.

georgeclaghorn commented on actiontext/lib/action_text/railtie.rb:
```diff
@@ -0,0 +1,13 @@
+  class Railtie < Rails::Railtie # :nodoc:
+    config.action_text = ActiveSupport::OrderedOptions.new
+    config.action_text.attachment_tag_name = "action-text-attachment"
+    initializer "action_text.configure" do |app|
```
Blank line before `initializer`, please.

georgeclaghorn commented on actiontext/test/unit/content_test.rb:
```diff
@@ -131,4 +140,14 @@ def content_from_html(html)
+    def reset_attachment_tag_name
+      ActionText::Attachment.tag_name = "action-text-attachment"
+      ActionText::Attachment.selector = "action-text-attachment"
+    end
```
Block form with automatic reset is idiomatic here:

```ruby
def with_attachment_tag_name(tag_name)
  previous_tag_name = ActionText::Attachment.tag_name
  ActionText::Attachment.tag_name = tag_name
  
  yield
ensure
  ActionText::Attachment.tag_name = previous_tag_name
end
```

georgeclaghorn commented on guides/source/configuring.md:
```diff
@@ -1121,6 +1121,11 @@ text/javascript image/svg+xml application/postscript application/x-shockwave-fla
+
+* `config.action_text.attachment_tag_name` accepts a string for the HTML tag used to wrap attachments. Defaults to 'action-text-attachment'.
+
+
```
Let’s move this up above “Results of `config.load_defaults`” with the rest of the framework components.

markvanlan commented:
@georgeclaghorn Thanks for the review. I removed `ActionText::Attachement.selector` and replaced it with `tag_name` where it was used. Addressed the rest of you review as well.

georgeclaghorn commented on actiontext/lib/action_text/railtie.rb:
```diff
@@ -0,0 +1,14 @@
+      ActionText::Attachment.tag_name = app.config.action_text.attachment_tag_name
+    end
+  end
+end
```
Just noticed that this is a new railtie, separate from the existing `ActionText::Engine`. Can we fold this into `ActionText::Engine`, or was this intentional?

markvanlan replied:
I just tried to follow patterns looking at other parts of the codebase. I can fold it into `ActionText::Engine`!

markvanlan commented:
@georgeclaghorn I removed the railtie. I'm not sure why all these specs are failing, they look flakey or unrelated? For example `/railties/test/generators/app_generator_test.rb` passes locally but [not in CI](https://buildkite.com/rails/rails/builds/75561#8405ee88-205b-4ec2-99f0-5f866c506e72) 

```
ws/rails/railties  actiontext-config-attachment-tag-name ✔              16m
▶ bundle exec ruby -w -Itest test/generators/app_generator_test.rb
/home/markvanlan/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/mail-2.7.1/lib/mail/parsers/address_lists_parser.rb:32454: warning: statement not reached
/home/markvanlan/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/mail-2.7.1/lib/mail/parsers/address_lists_parser.rb:32615: warning: statement not reached
/home/markvanlan/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/mail-2.7.1/lib/mail/parsers/address_lists_parser.rb:32651: warning: statement not reached
/home/markvanlan/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/mail-2.7.1/lib/mail/parsers/address_lists_parser.rb:32740: warning: statement not reached
/home/markvanlan/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/mail-2.7.1/lib/mail/parsers/address_lists_parser.rb:32756: warning: statement not reached
/home/markvanlan/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/mail-2.7.1/lib/mail/parsers/address_lists_parser.rb:32822: warning: statement not reached
/home/markvanlan/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/mail-2.7.1/lib/mail/parsers/address_lists_parser.rb:32863: warning: statement not reached
/home/markvanlan/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/mail-2.7.1/lib/mail/parsers/address_lists_parser.rb:32888: warning: statement not reached
/home/markvanlan/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/mail-2.7.1/lib/mail/parsers/address_lists_parser.rb:31984: warning: assigned but unused variable - testEof
Run options: --seed 31500

# Running:
...............................................................................................................................
....
Finished in 11.534206s, 11.3575 runs/s, 88.2592 assertions/s.
131 runs, 1018 assertions, 0 failures, 0 errors, 0 skips
```

georgeclaghorn commented:
Could you merge main or rebase against it, please? I think that’ll fix the tests.

markvanlan commented:
🎉 Good call

georgeclaghorn commented:
Okay, last thing, I think: could you please add a changelog entry?

markvanlan commented:
I think I did it correctly 🤷

georgeclaghorn commented on actiontext/CHANGELOG.md:
```diff
@@ -10,5 +10,9 @@
+*  Add `config.action_text.attachment_tag_name`, to specify the HTML tag that contains attachments.
+
+    *Mark VanLandingham*
+
```
New changelog entries go on top.

markvanlan replied:
yep yep yep.
